### PR TITLE
Face neighbor data bugfix [bugfix-face-restriction]

### PR DESCRIPTION
**Pull request by jamiebramwell**
When using the `H1FaceRestriction` operator in parallel, it assumes the face neighbor data in the mesh has already been populated. If this has not happened, the face restriction operator constructs without error and erroneously includes shared faces. This fix ensures these data structures are made prior to the construction of the face restriction operator.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2399](https://github.com/mfem/mfem/pull/2399) | @jamiebramwell | @tzanio | @YohannDudouit + @v-dobrev + @pazner | 07/09/21 | 08/03/21 | 08/06/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/restriction.cpp [hunk 1]
```diff
@@ -684,6 +690,20 @@ H1FaceRestriction::H1FaceRestriction(const FiniteElementSpace &fes,
...
+#ifdef MFEM_USE_MPI
+
+   // If the underlying finite element space is parallel, ensure the face neighbor information is generated.
+   const ParFiniteElementSpace *pfes = dynamic_cast<const ParFiniteElementSpace*>
+                                       (&fes);
+
+   if (pfes != NULL)
+   {
+      pfes->GetParMesh()->ExchangeFaceNbrData();
+   }
```
Comment (pazner) [inline fem/restriction.cpp:703]:
```suggestion
   // If the underlying finite element space is parallel, ensure the face
   // neighbor information is generated.
   if (const ParFiniteElementSpace *pfes
       = dynamic_cast<const ParFiniteElementSpace*>(&fes))
   {
      pfes->GetParMesh()->ExchangeFaceNbrData();
   }
```

Minor nitpicks: wrapping comment at 80 columns, restricting scope of `pfes` to the `if` statement.

**Review by pazner — APPROVED** [review]
Good bug fix, thanks @jamiebramwell!

File: fem/restriction.cpp (hunk 1 above)
Comment (jamiebramwell) [reply fem/restriction.cpp:703]: I admit, I'm lazy and do whatever the style checker tells me to do. 😄

**Comment by adrienbernede** [comment]
LGTM

**Comment by tzanio** [comment]
@YohannDudouit and @v-dobrev -- the reviews for this PR are due

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
May need to merge `master` here to get clean CI results.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**